fdbclient/include/fdbclient/CommitTransaction.h

```diff
@@ -123,59 +129,123 @@ struct MutationRef {
 
 	std::string toString() const {
 		std::string checksumStr;
-		if (checksum.present()) {
-			checksumStr = format("checksum: %s ", std::to_string(checksum.get()).c_str());
+		std::string accumulativeChecksumIndexStr;
+		if (this->checksum.present()) {
+			checksumStr = format("checksum: %s ", std::to_string(this->checksum.get()).c_str());
+			if (this->accumulativeChecksumIndex.present()) {
+				accumulativeChecksumIndexStr = format("accumulativechecksumindex: %s ",
+				                                      std::to_string(this->accumulativeChecksumIndex.get()).c_str());
+			}
 		}
-		uint8_t cType = type & ~CHECKSUM_FLAG_MASK;
-		return format("%scode: %s param1: %s param2: %s",
+		uint8_t mType = this->type & ~CHECKSUM_FLAG_MASK;
+		mType = mType & ~ACCUMULATIVE_CHECKSUM_INDEX_FLAG_MASK;
+		return format("%s%scode: %s param1: %s param2: %s",
 		              checksumStr.c_str(),
-		              cType < MutationRef::MAX_ATOMIC_OP ? typeString[(int)cType] : "Unset",
+		              accumulativeChecksumIndexStr.c_str(),
+		              mType < MutationRef::MAX_ATOMIC_OP ? typeString[(int)mType] : "Unset",
 		              printable(param1).c_str(),
 		              printable(param2).c_str());
 	}
 
-	uint8_t typeWithChecksum() const { return this->type | CHECKSUM_FLAG_MASK; }
-	Optional<uint32_t> removeChecksum() {
-		this->checksum.reset();
-		if (!withChecksum()) {
-			return Optional<uint32_t>();
+	// Return true if the mutation param2 carries checksum
+	bool withChecksum() const { return this->type & CHECKSUM_FLAG_MASK; }
+
+	// Return true if the mutation param2 carries acs index
+	bool withAccumulativeChecksumIndex() const { return this->type & ACCUMULATIVE_CHECKSUM_INDEX_FLAG_MASK; }
+
+	// Set checksum bit to mutation type
+	uint8_t setTypeWithChecksum(uint8_t inType) const { return inType | CHECKSUM_FLAG_MASK; }
+
+	// Set acs bit to mutation type
+	uint8_t setTypeWithAccumulativeChecksumIndex(uint8_t inType) const {
+		return inType | ACCUMULATIVE_CHECKSUM_INDEX_FLAG_MASK;
+	}
+
+	bool isAtomicOp() const { return (ATOMIC_MASK & (1 << type)) != 0; }
+	bool isValid() const {
+		return type < MAX_ATOMIC_OP && type < 64U;
```

Comment:
Do we need the `type < 64U` part, since `MAX_ATOMIC_OP` is always less than 64

Reply:
Use `static_assert` for compiling time check.